### ISSUE TITLE
Loading pre-computed distance transform during training

### ISSUE THREAD

**Issue by andy-s-ding** [MIC-DKFZ/nnUNet#1028]
I've been recently working on distance-penalized loss functions for multi-class segmentation with nnUNet. The computational demand is high, which makes training significantly longer. Rather than computing the Euclidean distance transform (edt) for each class naively, there is a method (https://github.com/seung-lab/euclidean-distance-transform-3d) that handles multi-class edt calculations in two passes. Still, training can take a very long time (almost 2 days per fold for 16 classes in 512x512x512 CT images).

One strategy is to compute the edt of each training image offline and load it concurrently with the training data in the data loader. Since patches of the training data are loaded for training, I'm not really quite sure how to implement this in the best way to also load in the patch of the distance transforms for each class. Any thoughts? Thanks for your time!

**Comment by chr-wei** [comment]
@andy-s-ding Did you find an approach finally? I wanted to try something similiar. Probably you are not working into that direction anymore because the initial post was 2018.

**Comment by andy-s-ding** [comment]
@chr-wei I haven't found an approach yet. I've stayed with the edt method linked in my original post for simplicity. Apart from changing the data loading step to include the patch of the distance transform, I realize that data augmentation would theoretically change the values of the distance transform. Therefore, the distance transform patch initially loaded in may not accurately represent the distance map once it is transformed by data augmentation.

**Comment by FabianIsensee** [comment]
Have you seen this branch? 
https://github.com/MIC-DKFZ/nnUNet/tree/distance_maps

It is horribly outdated but it does exactly that. This branch was initially made for @JunMa11, I am not sure whether he is sill using it.
His repository also has some EDT-based losses: https://github.com/JunMa11/SegLoss/tree/master/losses_pytorch
He might be able to help
